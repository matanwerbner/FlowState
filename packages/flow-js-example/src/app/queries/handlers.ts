import { rest } from 'msw';
import { QuestionApiResponse } from '../lib/types';

const mockedSuccessResponse: QuestionApiResponse = {
  response_code: 1,
  results: [
    {
      category: 'Entertainment: Video Games',
      type: 'boolean',
      difficulty: 'medium',
      question:
        "The retail disc of Tony Hawk's Pro Skater 5 only came out in North America.",
      correct_answer: 'True',
      incorrect_answers: ['False'],
    },
  ],
};

export const httpHandlers = [
  rest.get('https://opentdb.com/api.php', (req, res, ctx) => {
    return res(ctx.status(200, 'success'), ctx.json(mockedSuccessResponse));
  }),
];
