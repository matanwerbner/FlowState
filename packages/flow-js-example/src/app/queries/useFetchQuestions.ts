import { useQuery } from '@tanstack/react-query';
import { GameContext } from '../lib/types';

export const useFetchQuestions = (
  context: Pick<GameContext, 'category' | 'difficulty'>
) =>
  useQuery({
    queryKey: ['FetchQuestions'],
    networkMode: 'always',
    queryFn: () => {
      const queryParams = new URLSearchParams({
        amount: '10',
        type: 'boolean',
      });
      if (context.category !== '') {
        queryParams.append('category', context.category);
      }
      if (context.difficulty !== '') {
        queryParams.append('difficulty', context.difficulty);
      }

      return fetch(`https://opentdb.com/api.php?${queryParams.toString()}`);
    },
    gcTime: 0,
  });
