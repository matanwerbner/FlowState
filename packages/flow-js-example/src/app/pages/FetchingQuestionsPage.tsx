import { useGameStore } from '../store';
import { useFetchQuestions } from '../queries/useFetchQuestions';
import { useEffect, useMemo, useState } from 'react';
import { QuestionApiResponse } from '../lib/types';
import { Loading } from '../components/Loading';
import { FlexBox } from '../components/FlexBox';
import { Button } from '@mui/material';

export const FetchingQuestionsPage = () => {
  const { context, triggerEvent } = useGameStore((state) => state);
  const { data, isLoading, isError, refetch } = useFetchQuestions(context);
  const [isEmpty, setIsEmpty] = useState(false);

  useEffect(() => {
    if (data && !isLoading && !isError) {
      data.json().then(({ results }: QuestionApiResponse) => {
        if (results.length === 0) {
          setIsEmpty(true);
        } else {
          triggerEvent('questionsFetched', {
            questions: results.map((question) => ({
              question: question.question,
              answer: question.correct_answer === 'True',
            })),
          });
        }
      });
    }
  }, [data, isError, isLoading, triggerEvent]);

  const content = useMemo(() => {
    if (isEmpty) {
      return (
        <FlexBox data-testid="emptyQuestionsContainer">
          <div>No questions found for this category and difficulty</div>
          <Button onClick={() => triggerEvent('playAgain')}>Start Over</Button>
        </FlexBox>
      );
    }
    if (isError) {
      return (
        <FlexBox>
          <div>Failed to fetch questions, </div>
          <Button onClick={() => refetch()}>Try again!</Button>
        </FlexBox>
      );
    }

    return <Loading />;
  }, [isEmpty, isError, refetch, triggerEvent]);

  return content;
};
