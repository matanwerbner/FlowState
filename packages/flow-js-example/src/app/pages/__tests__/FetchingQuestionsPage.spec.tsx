/* eslint-disable @typescript-eslint/ban-ts-comment */

import { render, waitFor, screen } from '@testing-library/react';
import { FetchingQuestionsPage } from '../FetchingQuestionsPage';
import * as fetchQuestions from '../../queries/useFetchQuestions';

describe('FetchingQuestionsPage tests', () => {
  it('should render error page if error', () => {
    const spy = jest.spyOn(fetchQuestions, 'useFetchQuestions');
    // @ts-expect-error
    spy.mockReturnValue({ isError: true });

    const { baseElement } = render(<FetchingQuestionsPage />);
    expect(baseElement).toMatchSnapshot();
  });

  it('should render empty page if empy', async () => {
    const spy = jest.spyOn(fetchQuestions, 'useFetchQuestions');
    spy.mockReturnValue({
      // @ts-expect-error
      data: { json: () => Promise.resolve({ results: [] }) },
    });

    render(<FetchingQuestionsPage />);

    await waitFor(() => {
      expect(screen.getByTestId('emptyQuestionsContainer')).toBeInTheDocument();
    });
  });

  it('should render loader', async () => {
    const spy = jest.spyOn(fetchQuestions, 'useFetchQuestions');

    // @ts-expect-error
    spy.mockReturnValue({
      isLoading: true,
    });

    const { baseElement } = render(<FetchingQuestionsPage />);

    expect(baseElement).toMatchSnapshot();
  });
});
