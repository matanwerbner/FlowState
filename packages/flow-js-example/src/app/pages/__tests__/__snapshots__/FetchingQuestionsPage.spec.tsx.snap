// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FetchingQuestionsPage tests should render error page if error 1`] = `
<body>
  <div>
    <div
      class="FlexBox__StyledBox-txfyk3-0 lgdOeB MuiBox-root css-329zn9"
    >
      <div>
        Failed to fetch questions, 
      </div>
      <button
        class="MuiButtonBase-root MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary MuiButton-root MuiButton-text MuiButton-textPrimary MuiButton-sizeMedium MuiButton-textSizeMedium MuiButton-colorPrimary css-1e6y48t-MuiButtonBase-root-MuiButton-root"
        tabindex="0"
        type="button"
      >
        Try again!
        <span
          class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
        />
      </button>
    </div>
  </div>
</body>
`;

exports[`FetchingQuestionsPage tests should render loader 1`] = `
<body>
  <div>
    <div
      class="Loading__StyledLoading-sc-1hxy68f-0 erMhjE"
      data-testid="loader"
    >
      <img
        alt="loading"
        src="loading.gif"
      />
    </div>
  </div>
</body>
`;
